lib/core/test/unit/Cardano/Wallet/DB/Store/Meta/StoreSpec.hs

```diff
@@ -0,0 +1,336 @@
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Cardano.Wallet.DB.Store.Meta.StoreSpec(
+    spec, deltasForManipulate
+) where
+
+import Prelude
+
+import Cardano.Wallet.DB.Arbitrary
+    ()
+import Cardano.Wallet.DB.Fixtures
+    ( WalletProperty
+    , coverM
+    , elementsOrArbitrary
+    , frequencySuchThat
+    , logScale
+    , unsafeUpdateS
+    , withDBInMemory
+    , withInitializedWalletProp
+    )
+import Cardano.Wallet.DB.Sqlite.Schema
+    ( TxMeta (txMetaDirection, txMetaSlot, txMetaStatus)
+    , txMetaSlotExpires
+    , txMetaTxId
+    )
+import Cardano.Wallet.DB.Sqlite.Types
+    ( TxId (TxId) )
+import Cardano.Wallet.DB.Store.Meta.Model
+    ( DeltaTxMetaHistory (..)
+    , ManipulateTxMetaHistory (..)
+    , TxMetaHistory (..)
+    , mkTxMetaHistory
+    , overTxMetaHistory
+    )
+import Cardano.Wallet.DB.Store.Meta.Store
+    ( mkStoreTransactionsMeta )
+import Cardano.Wallet.Primitive.Types
+    ( WalletId )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( Direction (Incoming, Outgoing), TxStatus (InLedger, Pending), status )
+import Control.Arrow
+    ( (***) )
+import Control.Monad
+    ( foldM )
+import Data.DBVar
+    ( Store )
+import Data.Foldable
+    ( toList )
+import Data.Map.Strict
+    ( Map )
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+    ( catMaybes, fromJust )
+import Data.Set
+    ( Set )
+import Database.Persist.Sql
+    ( SqlPersistT )
+import Test.DBVar
+    ( prop_StoreUpdates )
+import Test.Hspec
+    ( Spec, around, describe, it )
+import Test.QuickCheck
+    ( Gen, Property, arbitrary, elements, frequency, listOf1, property )
+import Test.QuickCheck.Monadic
+    ( PropertyM, assert, pick )
+
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Cardano.Wallet.Primitive.Types.Tx as W
+import qualified Data.Set as Set
+
+spec :: Spec
+spec = around withDBInMemory $ do
+    describe "Meta transactions store" $ do
+        it "respects store laws"
+            $ property . prop_StoreMetaLaws
+        it "can prune all not-in-ledger transaction"
+            $ property . prop_PruneToAllInLedger
+        it "can mark pending transactions as expired based on current slot"
+            $ property . prop_AgeAllPending2Expire
+        it "can mark past pending transactions as expired based on current slot"
+            $ property . prop_AgeSomePending2Expire
+        it "can roll back to a given slot, removing all incoming transactions"
+            $ property . prop_RollBackRemoveAfterSlot
+        it "can roll back to a given slot, switching all outgoing transactions \
+            \ after slot in pending state"
+            $ property . prop_RollBackSwitchOutgoing
+        it "can roll back to a given slot, preserving all outgoing transactions"
+            $ property . prop_RollBackPreserveOutgoing
+        it "can roll back to a given slot, leaving past untouched"
+            $ property . prop_RollBackDoNotTouchPast
```

Comment:
Question: Are these properties mainly about the `Store`, or are these properties mainly about the verbs and how they `apply` to `TxMetaHistory`? 🤔 In the latter case, I would suggest to make them pure and put them into a module `ModelSpec` instead.

I'm asking because the idea behind the `Store` concept is quite ruthless: As soon as we know that the stored values perfectly reflects the in-memory value (as asserted by `prop_StoreMetaLaws`), we *know* that the stored value behaves just as the in-memory value, and do not need to the behavior of `apply` separately. 🤓

(Of course, the caveat is that the statistics of the randomized testing in `prop_StoreMetaLaws` may not be sufficient to check that the "mirror" is spotless. But that is one reason why I'm in favor of making the implementation of `update` and `apply` very similar, using e.g. the same names etc.)

Reply:
I do not agree here. Apply is an implementation detail for me, a user of the store, I prefer to have the store with tests than the `apply`.

The preference is coming from a slightly different take on the value of tests.

Good tests IMO should: 

1. Prove the correctness of any implementation. In this case, a higher DSL is good so that we can change implementation by rewriting the interpreter of the tests only.
3. Document the specification. In this case by chance a higher DSL is good, so implementation doesn't surface and litter the semantic that should stand out test definition)

Here ultimately we care that the store object is correct and that I can understand its semantic reading the always-up-to-date documentation that the tests provide.

All your considerations are correct btw. We could test the implementation + the store consistency and have a ModelSpec module instead. I will give it a try.

Reply:
I implemented the ModelSpec, better, simpler code. I'll consider this a win.

Reply:
> I implemented the ModelSpec, better, simpler code. I'll consider this a win.

I agree. 😊

Welcome to the anti-monad resistance! ✊